### PR TITLE
Add windows support to fb_ntp

### PR DESCRIPTION
**Pull request by jaymzh**
This PR adds basic windows support to fb_ntp.

The support is just for setting NTP servers and ensureing the service
is running. There are other things one can set and this could be
extended if desired, but it works well for now.

Note this uses powershell_exec which I think requires Chef 16. It could
be moved to powershell_out if necessary, though that is much, *much*
slower.

This also has a chefstyle autocorrect which does implicit rescue which
requires Chef 13, but I _think_ the last of 12 is dead at FB so that
should be fine, but someone should doublecheck.

### CODE REVIEW

File: cookbooks/fb_ntp/resources/windows_config.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  want = node['fb_ntp']['servers']
+  have = config['TimeProviders']['NtpServer'].split(',')
+  if Set.new(want) != Set.new(have)
+    Chef::Log.info(
+      'fb_ntp_windows_config: Changing NTP servers from ' +
+      "#{have.join(', ')} to #{want.join(', ')}",
+    )
+    set_ntp_servers
+  else
+    Chef::Log.debug('fb_ntp_windows_config: NTP servers are correct')
```
Comment (davide125) [inline cookbooks/fb_ntp/resources/windows_config.rb:37]: nit: the convention for logging in LWRPs is to use `fb_ntp[windows_config]`

File: cookbooks/fb_ntp/resources/windows_config.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
+action_class do
+  def get_current_config
+    config = {}
+    section = nil
+    s = powershell_exec('w32tm /query /configuration')
```
Comment (davide125) [inline cookbooks/fb_ntp/resources/windows_config.rb:5]: can/should this be moved to a ohai plugin?

**Comment by davide125** [comment]
You'll also need to update `metadata.rb`

File: cookbooks/fb_ntp/resources/windows_config.rb (hunk 2 above)
Comment (jaymzh) [reply cookbooks/fb_ntp/resources/windows_config.rb:5]: No, resources should always check the state at runtime.

File: cookbooks/fb_ntp/resources/windows_config.rb (hunk 1 above)
Comment (jaymzh) [reply cookbooks/fb_ntp/resources/windows_config.rb:37]: yes, good call.

File: cookbooks/fb_ntp/recipes/default.rb [hunk 3]
```diff
@@ -43,16 +47,14 @@
...
-      rescue Resolv::ResolvError
-        Chef::Log.warn("fb_ntp: failed to resolve #{host}, skipping")
+      ips = Resolv.getaddresses(host)
+      ips.each do |ip|
+        dash6 = ''
+        dash6 = '-6 ' if IPAddr.new(ip).ipv6?
+        acl_entries << "restrict #{dash6}#{ip} ##{host}"
       end
+    rescue Resolv::ResolvError
+      Chef::Log.warn("fb_ntp: failed to resolve #{host}, skipping")
```
Comment (malmond77) [inline cookbooks/fb_ntp/recipes/default.rb:57]: This changes the error handling to skip single servers that don't resolve, to skipping all servers past the first one that fails. Why the change in semantics here?

Comment (jaymzh) [reply cookbooks/fb_ntp/recipes/default.rb:57]: rubocop did this automatically. :( seems like a bug, you're right. :/

Comment (jaymzh) [reply cookbooks/fb_ntp/recipes/default.rb:55]: Nope, wait, you're misunderstanding the code. The implicit rescue there doesn't stop the loop... it rescues. The code is identical:

```
[phil@rider ~]$ cat /tmp/test.rb 
#!/usr/bin/ruby

(0..9).each do |i|
  puts i
  raise if i == 5
rescue
  puts 'i am at 5'
end
[phil@rider ~]$ /tmp/test.rb
0
1
2
3
4
5
i am at 5
6
7
8
9
```

That code is correct.

**Comment by jaymzh** [comment]
> You'll also need to update `metadata.rb`

for what? this doesn't have a supports.

**Comment by jaymzh** [comment]
should be good now @davide125

File: cookbooks/fb_ntp/recipes/default.rb (hunk 3 above)
Comment (malmond77) [reply cookbooks/fb_ntp/recipes/default.rb:55]: This sort of thing seems to be par for the course for Ruby. Alright, TIL then.

**Review by davide125 — APPROVED** [review]

**Pull request closed without merging**